### PR TITLE
Core/Spells now the spells for summon mount aren't affected by haste

### LINKED ISSUE #554: Haste and mounts
**Description**: Class specific mounts are bugged. The cast is faster with high amount of haste, as it was a normal spell.

**Current behaviour**: Currently summoning your mount is faster if you have more haste.

**Expected behaviour**: The cast should be fixed at 1.5 sec!

**Steps to reproduce the problem**:

Shut up a paladin, and note, for example, that the cast of the Flash of Light (which is 1.5 basic, and is rightly speeded up) is the same as that of the mount.

**Branch(es)**: master

**AC hash/commit**:  6aff74cd1760b4fd170f8869cc625741e83a9b23

**Operating system**:  Ubuntu 16.04

**Modules**:  None

**Other Customizations**: None

### PR DESCRIPTION
**Pull request by Ercules76**
**Target branch(es):** Master.

**Issues addressed:** Closes #554 

This PR resolve the issue https://github.com/azerothcore/azerothcore-wotlk/issues/554

### CODE REVIEW

File: src/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -3759,7 +3759,7 @@ void Spell::handle_immediate()
             if (Player* modOwner = m_caster->GetSpellModOwner())
                 modOwner->ApplySpellMod(m_spellInfo->Id, SPELLMOD_DURATION, duration);
             // Apply haste mods
-            if (m_caster->HasAuraTypeWithAffectMask(SPELL_AURA_PERIODIC_HASTE, m_spellInfo) || m_spellInfo->HasAttribute(SPELL_ATTR5_HASTE_AFFECT_DURATION))
+			if (m_caster->HasAuraTypeWithAffectMask(SPELL_AURA_PERIODIC_HASTE, m_spellInfo) || m_spellInfo->HasAttribute(SPELL_ATTR5_HASTE_AFFECT_DURATION))
```
Comment (talamortis) [inline src/game/Spells/Spell.cpp]: Please fix indentation.

File: src/game/Entities/Unit/Unit.cpp [hunk 2]
```diff
@@ -13634,15 +13634,22 @@ int32 Unit::ModSpellDuration(SpellInfo const* spellProto, Unit const* target, in
     }
     return std::max(duration, 0);
 }
-
-void Unit::ModSpellCastTime(SpellInfo const* spellProto, int32 & castTime, Spell* spell)
+// [AZTH] fix by MadAgos
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: remove this please

File: src/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -13634,15 +13634,22 @@ int32 Unit::ModSpellDuration(SpellInfo const* spellProto, Unit const* target, in
...
+void Unit::ModSpellCastTime(SpellInfo const* spellInfo, int32 & castTime, Spell* spell)
 { 
-    if (!spellProto || castTime < 0)
+    if (!spellInfo || castTime < 0)
         return;
+
+	if (spellInfo->IsChanneled() && !(spellInfo->AttributesEx5 & SPELL_ATTR5_HASTE_AFFECT_DURATION))
+		return;
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: indentation

File: src/game/Entities/Unit/Unit.cpp [hunk 4]
```diff
@@ -13634,15 +13634,22 @@ int32 Unit::ModSpellDuration(SpellInfo const* spellProto, Unit const* target, in
...
+
     // called from caster
     if (Player* modOwner = GetSpellModOwner())
-        modOwner->ApplySpellMod(spellProto->Id, SPELLMOD_CASTING_TIME, castTime, spell, bool(modOwner != this && !IsPet()));
 
+		// TODO: (MadAgos) eventualmente controllare ed eliminare l'argomento bool
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: only english comments plz :D

File: src/game/Entities/Unit/Unit.cpp [hunk 5]
```diff
@@ -13654,7 +13661,31 @@ void Unit::ModSpellCastTime(SpellInfo const* spellProto, int32 & castTime, Spell
         castTime = int32(float(castTime) * m_modAttackSpeedPct[RANGED_ATTACK]);
     else if (spellProto->SpellVisual[0] == 3881 && HasAura(67556)) // cooking with Chef Hat.
         castTime = 500;
-}
+		*/
+
+
+	switch (spellInfo->DmgClass)
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: ?

File: src/game/Entities/Unit/Unit.cpp [hunk 6]
```diff
@@ -13654,7 +13661,31 @@ void Unit::ModSpellCastTime(SpellInfo const* spellProto, int32 & castTime, Spell
...
+		break;
+	case SPELL_DAMAGE_CLASS_RANGED:
+		castTime = int32(float(castTime) * m_modAttackSpeedPct[RANGED_ATTACK]);
+		break;
+	default:
+		break;
+	}
+}
+
+// [/AZTH]
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: this is not supposed to be here

File: src/game/Spells/Spell.cpp (hunk 1 above)
Comment (Ercules76) [inline src/game/Spells/Spell.cpp]: correct

File: src/game/Entities/Unit/Unit.cpp (hunk 2 above)
Comment (Ercules76) [inline src/game/Entities/Unit/Unit.cpp]: I remove the AZTH tags

File: src/game/Entities/Unit/Unit.cpp (hunk 4 above)
Comment (Ercules76) [inline src/game/Entities/Unit/Unit.cpp]: done

File: src/game/Entities/Unit/Unit.cpp [hunk 7]
```diff
@@ -13635,25 +13635,38 @@ int32 Unit::ModSpellDuration(SpellInfo const* spellProto, Unit const* target, in
...
+    if (!spellInfo || castTime < 0)
+        return;
+
+    if (spellInfo->IsChanneled() && !(spellInfo->AttributesEx5 & SPELL_ATTR5_HASTE_AFFECT_DURATION))
         return;
+
     // called from caster
     if (Player* modOwner = GetSpellModOwner())
-        modOwner->ApplySpellMod(spellProto->Id, SPELLMOD_CASTING_TIME, castTime, spell, bool(modOwner != this && !IsPet()));
+        // TODO:(MadAgos) Eventually check and delete the bool argument
```
Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: please resolve all TODOs before merging

Comment (ShinDarth) [inline src/game/Entities/Unit/Unit.cpp]: ( or remove them or change them into a more descriptive note that can be used for reference in future )

**Pull request merged**